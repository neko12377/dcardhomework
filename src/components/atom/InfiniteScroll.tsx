import React, { useCallback, useEffect, useRef, useState } from "react";
import {
    PostContentInterface,
    useDataGetting,
} from "../../hooks/useDatsGetting";
import styled from "@emotion/styled";

const Base = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: center;
    background-color: white;
    width: 50%;
    overflow-y: auto;
    height: 100%;
    max-height: 100%;
`;

const Block = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;
    cursor: pointer;
    padding: 30px 10%;
    border-radius: 4px;
    width: 100%;

    &:hover {
        background-color: rgba(203, 203, 203, 0.45);
    }
`;

const Title = styled.h1`
    color: #2e2e2e;
    width: 100%;
`;

const Excerpt = styled.div`
    display: flex;
    justify-content: flex-start;
    align-items: flex-start;
    color: #3e3e3e;

    width: 100%;
`;

const Hr = styled.hr`
    width: 100%;
    margin-bottom: 20px;
    margin-top: 30px;
    border-bottom: 1px solid #a3a0a0;
`;

const Loading = styled.div`
    position: absolute;
    display: flex;
    top: 50%;
    left: 44%;
    background-color: rgba(151, 151, 151, 0.6);
    width: 150px;
    height: 80px;
    justify-content: center;
    align-items: center;
    border-radius: 8px;

    & div {
        display: flex;
        position: relative;
        width: 80px;
        height: 70px;
    }

    & div div {
        display: inline-block;
        position: absolute;
        left: 8px;
        width: 16px;
        background: rgba(255, 255, 255, 0.73);
        animation: lds-facebook 1.2s infinite;
    }

    & div div:nth-of-type(1) {
        left: 8px;
        animation-delay: -0.24s;
    }

    & div div:nth-of-type(2) {
        left: 32px;
        animation-delay: -0.12s;
    }

    & div div:nth-of-type(3) {
        left: 56px;
        animation-delay: 0s;
    }

    @keyframes lds-facebook {
        0% {
            top: 16px;
            height: 44px;
        }
        25% {
            top: 8px;
            height: 64px;
        }
        50%,
        100% {
            top: 24px;
            height: 32px;
        }
    }
`;

interface BackToTopButtonProps {
    visibility: "flex" | "none";
}

const BackToTopButton = styled.a<BackToTopButtonProps>`
    position: absolute;
    bottom: 20px;
    right: 26.5%;
    width: 50px;
    height: 50px;
    border-radius: 50%;
    background-color: rgba(56, 99, 158, 0.66);
    display: ${(props) => props.visibility};
    justify-content: center;
    align-items: center;
    text-decoration: none;
    color: white;
    font-weight: 900;
    cursor: pointer;

    &:hover {
        background-color: rgb(177, 231, 253);
    }
`;

const InfiniteScroll = () => {
    const [urlPath, setUrlPath] = useState<string>(
        "/proxy_domain_d_card/posts?popular=false"
    );
    const { posts, hasMore, isLoading, error } = useDataGetting(urlPath);
    const [currentLastId, setCurrentLastId] = useState<number | undefined>(
        posts[posts.length - 1]?.id
    );
    const postArray =
        posts.length > 0 && !error
            ? posts.map((item: PostContentInterface) => {
                  const { title, excerpt, id } = item;
                  return {
                      title,
                      excerpt,
                      id,
                  };
              })
            : [
                  {
                      title: "Currently not thing is available",
                      excerpt: "woo woo",
                  },
              ];

    const observer = useRef<null | IntersectionObserver>();
    const lastPostBlock = useCallback(
        (lastNode) => {
            observer.current && observer.current.disconnect();
            observer.current = new IntersectionObserver(
                (entries) => {
                    if (entries[0].isIntersecting) {
                        currentLastId &&
                            setUrlPath(
                                `/proxy_domain_d_card/posts?popular=false&before=${currentLastId}`
                            );
                    }
                },
                { threshold: 0.8 }
            );
            lastNode && observer.current?.observe(lastNode);
        },
        [currentLastId]
    );

    const [backToTopButtonVisibility, setBackToTopButtonVisibility] = useState<
        "flex" | "none"
    >("none");
    const topObserver = useRef<null | IntersectionObserver>();
    const firstPostBlock = useCallback((firstNode) => {
        topObserver.current = new IntersectionObserver((entries) => {
            !entries[0].isIntersecting && setBackToTopButtonVisibility("flex");
            entries[0].isIntersecting && setBackToTopButtonVisibility("none");
        });
        topObserver.current?.observe(firstNode);
    }, []);

    useEffect(() => {
        hasMore && setCurrentLastId(posts[posts.length - 1]?.id);
    }, [posts, hasMore]);
    return (
        <Base>
            {isLoading && (
                <Loading>
                    <div>
                        <div />
                        <div />
                        <div />
                    </div>
                </Loading>
            )}
            {postArray.map((item: PostContentInterface, index: number) =>
                postArray.length === index + 1 ? (
                    <Block key={`${item.title}_${item.id}`} ref={lastPostBlock}>
                        <Title>{item.title}</Title>
                        <Excerpt>{item.excerpt}</Excerpt>
                    </Block>
                ) : index === 0 ? (
                    <Block
                        key={`${item.title}_${item.id}`}
                        ref={firstPostBlock}
                        id="top"
                    >
                        <Title>{item.title}</Title>
                        <Excerpt>{item.excerpt}</Excerpt>
                        <Hr />
                    </Block>
                ) : (
                    <Block key={`${item.title}_${item.id}`}>
                        <Title>{item.title}</Title>
                        <Excerpt>{item.excerpt}</Excerpt>
                        <Hr />
                    </Block>
                )
            )}
            <BackToTopButton visibility={backToTopButtonVisibility} href="#top">
                TOP
            </BackToTopButton>
        </Base>
    );
};

export default InfiniteScroll;
