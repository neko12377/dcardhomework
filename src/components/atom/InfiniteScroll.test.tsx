/**
 * @jest-environment jsdom
 */

import React from "react";
import renderer from "react-test-renderer";
import InfiniteScroll from "./InfiniteScroll";

const intersectionObserverMock = () => ({
    observe: () => null,
    disconnect: () => null,
});
window.IntersectionObserver = jest
    .fn()
    .mockImplementation(intersectionObserverMock);

test("infinite scroll", () => {
    const component = renderer.create(<InfiniteScroll />);

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
});
