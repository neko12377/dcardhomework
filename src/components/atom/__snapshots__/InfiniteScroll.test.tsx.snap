// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`infinite scroll 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: white;
  width: 50%;
  overflow-y: auto;
  height: 100%;
  max-height: 100%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  padding: 30px 10%;
  border-radius: 4px;
  width: 100%;
}

.emotion-1:hover {
  background-color: rgba(203, 203, 203, 0.45);
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  font-size: 14px;
}

.emotion-3 {
  margin: 5px 0 10px 0;
  color: #2e2e2e;
  width: 100%;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  color: #3e3e3e;
  width: 100%;
}

.emotion-5 {
  position: absolute;
  bottom: 20px;
  right: 26.5%;
  width: 50px;
  height: 50px;
  border-radius: 50%;
  background-color: rgba(56, 99, 158, 0.66);
  display: none;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: white;
  font-weight: 900;
  cursor: pointer;
}

.emotion-5:hover {
  background-color: rgb(177, 231, 253);
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <div
      className="emotion-2"
    >
      男・undefined
    </div>
    <h1
      className="emotion-3"
    >
      Currently not thing is available
    </h1>
    <div
      className="emotion-4"
    >
      woo woo
    </div>
  </div>
  <a
    className="emotion-5"
    href="#top"
    visibility="none"
  >
    TOP
  </a>
</div>
`;
